core/configuration/src/main/kotlin/sqldelight/SQLDelightDrivers.kt

```diff
@@ -0,0 +1,18 @@
+package sqldelight
+
+import com.squareup.sqldelight.db.SqlDriver
+import com.squareup.sqldelight.sqlite.driver.JdbcSqliteDriver
+import java.io.File
+
+internal object SQLDelightDrivers {
+    private val instances: MutableMap<String, SqlDriver> = mutableMapOf()
+
+    fun instance(name: String): SqlDriver {
+        return instances[name] ?: synchronized(this) {
```

Comment:
synchronizing function is later called within coroutine. This is not recommended. Recommendation: Make function suspending or return a Deferred for the database.

Reply:
Where did you read that?

using `synchronize` inside a coroutine is indeed bad, but we aren't in a coroutine here.
This code isn't heavy / asynchronous either.

The synchronization here is just the pattern to make sure that you have a singleton.

Even if we returned a Deferred here it would not exempt you from the need of making sure you have a Singleton "Deferred" and thus you'd still have to use this double checked locking pattern.

see https://github.com/boswelja/android-dev-discord-bots/pull/3#discussion_r1039229438

If you ever looked inside `by lazy` by default it uses exactly this pattern to initialize the lazy, because it's the only safe way to initialize a singleton lazily.

The other option is to not do it lazily, but that means deciding the name of the database file in this module.

Reply:
I can see what the issue is here, but at the same time it's not doing significant work
@ThisIsIPat if you have a solution, would you be happy to open a PR once this one is in?

Reply:
I have previously encountered race conditions by calling lazy variables heavily within coroutines.

This is because the same coroutines dispatchers thread may be used twice during two parallel lazy calls. This is a general problem with synchronized within coroutines, as one thread will hold the lock but dispatched coroutines aren't thread-bound.

In this case, the instance() method is used within a coroutine (suspending function main()). While instance() is not suspending, it is still called from within a coroutine.

Reply:
Yes I can fix it too. If that's ok, feel free to close the convo.